[DirectX 12] Backend

wh1t3lord opened the pull request:
# Introduction

So it is a new backend from me and it's DirectX 12 for Microsoft NT platforms (>=Windows10).

The library comes to a stage that it is needed to provide more user friendly interaction between end user and library, what relates to rendering backend is a simple and naive integration like in the popular library ImGui by Ocornut where user already has an integrated backend and they need to pass an initialization structure with all required parameters. 

Well I will describe more detailed this chapter but for now I just leave this message in a such poor state, because for now it is a reminder for me to finish this work and move to Vulkan implementation. 

## Backend architecture

I set a straight aim to provide an efficient and optimized resource management and provide to user the low level settings of resource allocation. I mean that backend strives for being simple from user's side (in terms of how to set things up) and from another side it optimizes in all ways possible. 

### Buffer management aka geometry 

About buffer management the library allocates one buffer and utilizes all its space the offset calculation is based on this offset allocator implementation by sebbbi (see sources, MIT license).

But if buffer is out of memory or offset allocator can't allocate the required space it will allocate a new buffer. 

And rendering system tries to deallocate the buffer if it is not used. (in optimized way of course)

For now I tested with a really small sized buffers like less 1Kb and it was stable, of course it is not optimal settings due to high allocation of buffers and defragmentation of GPU's memory. By default I picked an optimal size that for most cases should be enough for using, but as I stated earlier user has a right to change that allocation size on initialization stage.

### Texture management

DirectX 12 has two types of allocation of textures. 

First one is to send texture to GPU as a single buffer, like if it is 4 Mbs then we send 4  Mbs as one single buffer. It is straight and simple and generally used among all modern GAPIs.

But there's a second variant for allocating textures aka placement PlacedResource (see [link](https://learn.microsoft.com/en-us/windows/win32/direct3d12/uploading-resources)). That means you can allocate multiple textures into one buffer like you allocate one big buffer and suballocate space for textures thus it is like you "placed" those textures into one buffer. But that works only for not big sized textures that less some specific size, because in official tests it won't give any profit from performance, but still useful for rational resource management. 

So based on texture size and its calculation the system tries to choose best way for uploading it. 

As a conclusion I tried to make backend that utilizes all memory efficiently as much as possible and make all possible to fill all buffers and buffers for textures in most optimal way that user doesn't need even think about it at all.

## Warning

Since we formally force people to use our backend then I don't see any motivation to help to people who make own backends based on our implementations and there's a vital reason for this, because it is better to have "one" (not in directly meaning) backend that be tested on many devices across different OSes than having some custom backends that can't be well tested at all. It is important warning for all users who have tendency to make own wheel ignoring obvious sane way of living. 

I need to say that I tried to keep things without breaking changes at all, like if you want to initialize RmlUI as you did before you just don't pass an init structure and from function calling it is just a default argument that you don't specify thus nothing is changed/broken.

But we will (at least me) accept all incoming that relate to backend and it's working state on different graphics cards and OSes for @mikke89 I suggest to create template issue that has the following form (just as an idea):

#### If you have trouble with rendering backend

#### Specify GAPI
Blah 12

#### Specify graphics card's driver
Blah 12.451.16 oct 2023

#### Specify graphics card 
Blah RTX 24000

#### Specify OS
Blah 13

#### Specify version and commit of  RmlIUI
6.0 d23a3dae6943e482fa1a2ed89c2dd48ccddc549f

I guess after finishing "modern" backends we will boost for some time the popularity of library but at the same time amount of running tests and examples would be high since everyone can easily integrate the library... :D

For now I am making support for multisampling feature and I hope I will finish this work till August.

mikke89 commented:
Thanks a lot for working on this, a DirectX 12 backend will be a very welcome addition! Let me know once you feel it is ready for review.

I quite like the initiative for introducing such a "renderer initialization" type that provides the user's own parameters. Although, I'll have to think a bit more about this, and how it fits into the library and how it integrates. I think it's important to keep it so that we don't take control over the main loop and event system, to ensure that the library plugs in to their existing infrastructure. Which is why we currently say that users should copy the backend itself into their own code (as opposed to the renderer and platform parts of the backend). But at the same time, we want to make it easier to get started for users that perhaps only want to provide their renderer details, but otherwise use the backend as it is. This generally looks like a nice opt-in solution that could be helpful in many use cases, but I'll need to take some more rounds and think about it.

Regarding issue templates. I actually quite like the informal nature of giving users a blank slate to describe their issues. I think that has some value as a community. Of course, I understand that with growing number of issues, some system like this could be helpful to speed things up, but I don't believe this is an issue for us at this point. I generally find that users are good at describing their issues. Regardless of templates, there will always be some times where more information is necessary. And that is perfectly fine in my view, there is also some value to this interaction as well. So I would prefer to keep this as it is.

Paril commented:
> Regarding issue templates. I actually quite like the informal nature of giving users a blank slate to describe their issues.

Just wanted to chime in on this one and say I agree; overly-complicated forms and templates have dissuaded me from making issues on other projects (same for something like stalebot, which I hate even more). I understand why some people use them but I find most people are pretty good with issue-posting when you don't give them any instructions at all.

wh1t3lord commented:
@mikke89 Hi and I have some good news! :smiley:

> I know it is kinda late and I was hell busy in that matter of time when I wrote that I am gonna finish that thing :disappointed: (I swear) but what happened that happened :grin: 

![image](https://github.com/user-attachments/assets/900a5858-431e-4c25-ae6c-a48ae41fbf23)

![image](https://github.com/user-attachments/assets/ef941949-728a-4a49-ba31-d09b74a83a01)

![image](https://github.com/user-attachments/assets/b5d6981b-816a-4130-8dfb-fcf4ccffaa15)

DirectX-12 probably is closing to be finished! Also I would like to add that I marked all methods for CPU and GPU profilers (GPU only those methods/functions where any calling from API is issued and it is helpful for debugging purposes for marking functions callings from API)

Anyway, I need to warn that you should think about passing invalid viewport/scissors and how it handle from frontend side because in DirectX-12 the scissor test is always enabled and in such case we get error like nothing happens much on Debug you will receive error (like debugbreak every time) or if you suppress (what I did now) a warning and it is kinda annoying :D 

I did experiment and treated a such situation where invalid scissor and vp is passed like not suppressing from API side but manually and in such case half of frames are not rendered at all like you get empty clear window's background without those buttons and other subwindows regions so it didn't work out as expected. 

The warnings are like these: 
> D3D12 WARNING: ID3D12CommandList::DrawIndexedInstanced: Viewport: 0 is non-empty while the corresponding scissor rectangle is empty.  Nothing will be written to the render target when this viewport is selected.  In D3D12, scissor testing is always enabled. [ EXECUTION WARNING #695: DRAW_EMPTY_SCISSOR_RECTANGLE]

> D3D12 WARNING: ID3D12CommandList::ClearRenderTargetView: Empty rect (left >= right or top >= bottom) means it not will clear anything.pRects[0] = { left:108, top:960, right:212, bottom:960 } [ EXECUTION WARNING #448: DEVICE_CLEARVIEW_EMPTYRECT]

Currently I need to finish some little things like checking Release build configuration (maybe I forgot something but geometry is kinda sus rendering but on Debug is all fine even validation layers didn't say anything illegal, no errors) and provide implementation of integration sample and to show minimal sample how RmlRenderInitInfo works (it is just matter of some hours, probably tomorrow I will write back here), and implement SDL, GLFW backend for DX12 it is small tasks too.

Because now we get everything works like all existed samples + effects sample but on your side I would be glad if you check visual tests and report something unusual or just some mistakes (probably when I will write that it is ready to be merged). I won't touch Readme and you can fill all needed information like updating backend table :)

I would like to hear any critique for default arguments that can be changed for/from user and they exist here

https://github.com/mikke89/RmlUi/blob/cbb4ea1015cc633e42fbe9906cb611da061cc498/Backends/RmlUi_Include_DirectX_12.h

And discuss it like what you like what you don't like and etc, because my aim was to handle all situations like why I choose preprocessors because some users can build RmlUI manually and they would like to change constants but themselves without CMake generations steps (configuration) from another hand there's a thing for CMake users too they can just pass their overrides and just use what they want to and if user doesn't need to change anything they will be used (and just used) as defaults in case if user didn't specify in RmlRenderInitInfo (I didn't make CMake side so it is better for someone to finish that feature, so it means that defaults will be generated and maybe as whole file through CMake)
 
About RmlRenderInitInfo it serves for two purposes
 
 1. Just pass all related things for using backend out-of-box, if it is detailed GAPI like DX12,VK they pass devices swapchains and other things
 2. If user specifies that it needs to override internal resource allocation policies, how much memory we reserve, how much of other things then that information will be use at initialization stage. A such feature is really useful among different teams from big projects or small they just pass their required limits for shipping and it works.
 
Also something worries me about constant buffers (uniform buffers) I mean they're a lot of them and would be better to think about possible optimization how to pass that data to shaders in reduced manner, maybe it is okay, but if there's something that might be improved in terms of their amount it would be perfect (I didn't think personally about how to reduce them yet)
 
I would like to hear other people and who used our backends and to hear their wishes/needs/discontent

wh1t3lord commented:
@mikke89 hi mikke I implemented the integration sample and I really recommend for you to implement the GL3 integration samples following my design and understanding my ideas and way of implementing things, but I tried to achieve two things: 

1. Make really small steps for integrating for user and thus easy callings, simplifying and being nice for consumer
2. Make things without any breaking changes related to current RmlUi samples and generally RmlUi functions of namespace Backend

But I just spend not so much time for making things beauty (in terms of code) and for that I recommend that thing for you :))

At least just to know that you would make things as you want to see because I can't know how's better to handle/write things from your side and not making any interferences you can finalize my idea about integration thing so better for you to refactor I just made things working.

Then I really want to upgrade our Vulkan and make effects sample working and we will have new way of using our backends for user environment and they don't need to copy/paste files and just directly use our code of RmlUi library. 😄 🎊 💯 

Yeah I will finish postprocess integration sample, but for now swapchain works it is just example of load document sample and I think I need to test on effects sample that will be renderered onto swapchain image.

So my idea is to provide for user two integration examples:

1) First example when user just wants to pass swapchain images directly where graphics information be rendered into swapchain
2) Second example when user just wants to render things as postprocess so he just passes single render target (framebuffer) renders rmlui interface onto that image and then using quad geometry rendering that postprocess image as texture for that quad and displaying on swapchain
3) This example just demonstrates same as postprocess but user wants to us make own command list and execute things so when Backend::EndFrame() is called on user's side he will get already rendered postprocess image without waiting and pushing commands for queue execution, this example is not added yet as well as a such feature as making isolated command list and command list execution is not finalized yet too 

So yeah, the main thing of integration sample it must be implemented from scratch demonstrating user engine's environment and show all things in one file and we just show using commentaries the needed steps for integration. I really want to see some users that have multithreaded renderers in their engines and thus they would like to render UI in some of separated thread and they have to report like there's no samples that will demonstrate a such integration, but still it must be a generalized solution, we will see I guess.

The really last thing what I need to fix and resolve for Dx12 it is CompositeLayers function where Blur and DropShadow render somehow the image result is flickering and idk why I forced to use Flush method (you can see in https://github.com/mikke89/RmlUi/blob/f96808d5df3a28a50d956954474ace05c19666c9/Backends/RmlUi_Renderer_DX12.cpp#L3405) and for some matter of time it is okay (performance doesn't degenerate) but still it is not right in terms of development and rendering, I will investigate and fix it for now just don't want to focus on that yet. 

I tested my backend on three videocards they are all Nvidia videocards so 

GTX 1060 Ti 6 GB (PC), GTX 1660 Ti 6 GB (PC),  RTX 3060 laptop 6 GB (notebook)

No one of them reported any problems, but I suggest to your known testers (and you personally) to use their videocards and run our samples on debug make with RMLUI_DX_DEBUG preprocessor enabled in order to get and receive any missed barrier (because when I thought that I finished development and when my 3060 didn't show any troubles the GTX 1660 showed me the one missed barrier so it is quite reasonable to test on other videocards that users have), BUT it is really important to run without Flush calling where I show in CompositeLayers.

Also funny fact the GTX 1060 renders well with or without Flush calling in CompositeLayers so I really hope it is possible to debug that thing and to understand where the problem on modern GPUs, but just FYI.

Currently integration samples will add only on DX12 (using standard procedure cmake -B Build_Win32_DX12 -S . --preset samples -DRMLUI_BACKEND=Win32_DX12 -DCMAKE_TOOLCHAIN_FILE="/<vcpkg-folder-path>/scripts/buildsystems/vcpkg.cmake"), but when you make GL3 you can add GL3 in cmake file of these samples by path Samples\basic\integration and Samples\basic\integration_2 (maybe I should rename as integration_swapchain and integration_2 as integration_postprocess but I thought it is too long names but whatever)

And yes, since we have a situation where backend and shell samples can run under different window management libraries or native that means we have to provide all integration samples for each of window management libraries or native implementation, for now we have only win32 I added temporary the others sdl and glfw for integration samples but I didn't complete them.

And yeah I know my explanation is kinda chaotic but it was needed to make RmlUi_Backend.cpp file where "generalized" functions will be called for user integration purposes (you will when you will try to implement on my pull GL3 samples).

The main question about design is where to store implementation functions for each backend where user requests in Backend::Initialize and other related functions that have similar design approach as Backend::Initialize version for integration (see rmlui_sample_integration_as_swapchain on dx12) because for now in RmlUi_Backend.h I made namespace for each backend and their functions and maybe we need to isolate from public space and add prefix ___ like I did for renderer_type variable, for now only dx12 implemented but storing in native files like Win32_DX12.cpp it is idk so better for you to think about where to store them maybe it is okay since we stored in RmlRenderInitInfo information about window handle and that window handle will be proceed in respected ProcessEvents based on window management library or native implementation again it is still better for you to think about how to make things nicer I just wrote dumb variables but obviously we have to have some internal structure and better accessing and all that stuff, so see variables ___renderer_ and p_integration_renderer).

Also we need to think about FileSystem and how to pass that thing for user, maybe we can use default ShellFileInterface but specify root from init info or just force user to pass own file interface (but I guess it is better that user expects that it resolves by itself like just hidden in Backend::Initialize function and maybe first approach is better where user just passes root maybe we can determine root if it wasn't passed by ourselves expecting just root as folder where application is executed or maybe to force user to pass root string otherwise if it wasn't passed we just say like failed to initialized backend)

https://github.com/mikke89/RmlUi/blob/4efd7cc77a2540acf38de640173f7b15f4c109a0/Backends/RmlUi_Backend_Win32_DX12.cpp#L274

Still I would like to hear any feedback from your side, thanks.

https://github.com/user-attachments/assets/0343611d-f9a7-40da-ab81-50100fd8af6a

wh1t3lord commented:
@mikke89 hi, any news? 🙃

mikke89 commented:
Hey, and sorry for the slow response, it's been a bit too busy for me lately. I will have some more time soon to delve into this, and get back to you. :)

wh1t3lord commented:
@mikke89 hi I pushed the fixes related to your last report and verify everything that works good please (and I didn't bring new bugs lol), for now I need to finish postprocess integration sample and that annoying blur bug that exists in effects sample.

Personally I hope when I port Vulkan it will show me exact problem where the sync primitive is missed but for now it is only iterational approach to understand what works not good due to missed barrier. It is def compositelayers place but debug validation doesn't say anything bad on all video cards 🫤🫤🫤

mikke89 commented:
Very nice! Thanks for the update, looks really good!

mikke89 commented:
That's great! Glad to hear that, and thanks for the continued work on this one.

I'm a bit backlogged right now, but I'll come around to this eventually. I might make some changes or additions (like the other backends) as I see fit, and then I'll likely just merge it. It's been some time in progress, and I agree that it's better to have something merged in sooner, rather then discussing minor details, and then everyone can help improve it over time.

Cheers!

wh1t3lord commented:
@mikke89 I fixed the issues :blush:

mikke89 commented:
That's really awesome, thanks! Good work :) I'll take one final glance over it and then expect to merge it soon.
